### PR TITLE
relatively small 71 file 700 lines changed refactor/rework to combat mode/sprint and some miscellaneous things that totally aren't important

### PR DESCRIPTION
**Pull request by silicons**
## About The Pull Request

also moved a bunch of booleans to new combat_flags
oh and status effects are processed by their own subsystem and have a 0.05 second resolution instead of 0.2.

major stuff:
combat mode lockout no longer turns it off it just disables its effects for the duration
same for sprint lockout somethign i added for future use with the stagger status effect

this means you can right click during combat mode lockout and toggle it.

## Why It's Good For The Game

more flexibility

also having your right click disabled by combat mode being off sucks, if this causes issues with right click interactions i can throw on combat mode checks rather than relying on right click BEING the check.

## Changelog
:cl:
refactor: combat mode/sprint/resisting lying down/attempting to crawl under all refactored, added combat/sprint mode lockouts, being locked out of combat mode no longer disables it and allows right click interactions etc etc
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/human/species.dm [hunk 1]
```diff
@@ -1794,9 +1788,12 @@ GLOBAL_LIST_EMPTY(roundstart_race_names)
 
 /datum/species/proc/althelp(mob/living/carbon/human/user, mob/living/carbon/human/target, datum/martial_art/attacker_style)
 	if(user == target && istype(user))
-		if(user.getStaminaLoss() >= STAMINA_SOFTCRIT)
+		if(IS_STAMCRIT(user))
 			to_chat(user, "<span class='warning'>You're too exhausted for that.</span>")
 			return
+		if(!IS_COMBAT_ACTIVE(user))
+			to_chat(user, "<span class='warning'>Your muscles need to be tensed to do that!</span>")
```
Comment (Linzolle) [inline code/modules/mob/living/carbon/human/species.dm]: I'm not sure if this is even possible for it to come up, but this seems kind of... non-descriptive? I can easily see how someone would be confused by this message if it ever came up.

File: code/modules/mob/living/living_combat.dm [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+/mob/living/proc/enable_combat_mode(silent = TRUE, was_forced = FALSE, visible = FALSE, update_icon = TRUE)
+	if(IS_COMBAT_ACTIVE(src))
+		return
+	ENABLE_BITFIELD(combat_flags, COMBAT_FLAG_COMBAT_ACTIVE)
+	SEND_SIGNAL(src, COMSIG_LIVING_COMBAT_ENABLED, was_forced)
+	if(visible)
+		visible_message("<span class='warning'>[src] drops into a combative stance!</span>", "<span class='warning'>You drop into a combative stance!</span>")
+	else if(!silent)
+		to_chat(src, was_forced? "<span class='warning'>Your muscles reflexively tignten!</span>" : "<span class='warning'>You tense your muscles.</span>")
```
Comment (Linzolle) [inline code/modules/mob/living/living_combat.dm]: tighten

File: code/modules/mob/living/carbon/human/species.dm (hunk 1 above)
Comment (silicons) [reply code/modules/mob/living/carbon/human/species.dm]: yeah alright and it comes up if you're combat mode locked but toggled on

File: code/modules/mob/living/living_combat.dm (hunk 2 above)
Comment (Linzolle) [reply code/modules/mob/living/living_combat.dm]: i was actually referring to the typo here, not to use a different word lol. "tignten"

File: code/__DEFINES/combat.dm [hunk 3]
```diff
@@ -29,7 +29,72 @@
...
+#define CAN_TOGGLE_COMBAT_MODE(mob)				((mob.stat == CONSCIOUS) && !IS_HARD_STAMCRITTED(mob))
+
+// Self documenting defines
+#define IS_RESISTING_REST(mob)				(mob.combat_flags & COMBAT_FLAG_RESISTING_REST)
+#define IS_INTENTIONALLY_RESTING(mob)		(mob.combat_flags & COMBAT_FLAG_INTENTIONALLY_RESTING)
+#define IS_ATTEMPTING_CRAWL(mob)			(mob.combat_flags & COMBAT_FLAG_ATTEMPTING_CRAWL)
+
+// Kind of bad file to put this in but it goes int othe same thing, I guess.
+#define IS_COMBAT_MODE_LOCKED(mob)			HAS_TRAIT(mob, TRAIT_COMBAT_MODE_LOCKED)
+#define IS_SPRINT_LOCKED(mob)				HAS_TRAIT(mob, TRAIT_SPRINT_LOCKED)
```
Comment (Ghommie) [inline code/__DEFINES/combat.dm]: Save for `CAN_TOGGLE_COMBAT_MODE`, all these macros do is simply obscure functionalities, which is not a good thing.

File: code/__DEFINES/combat.dm [hunk 4]
```diff
@@ -29,7 +29,72 @@
...
+#define COMBAT_FLAG_SOFT_STAMCRIT			(1<<8)
+
+// Helpers for getting someone's stamcrit state. Cast to living.
+#define NOT_STAMCRIT 0
+#define SOFT_STAMCRIT 1
+#define HARD_STAMCRIT 2
+
+// Stamcrit check helpers
+#define IS_HARD_STAMCRITTED(mob)			(mob.combat_flags & COMBAT_FLAG_HARD_STAMCRIT)
+#define IS_SOFT_STAMCRITTED(mob)			(mob.combat_flags & COMBAT_FLAG_SOFT_STAMCRIT)
```
Comment (Ghommie) [inline code/__DEFINES/combat.dm]: Same here.

File: code/modules/mob/living/living_combat.dm [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+/mob/living/proc/disable_combat_mode(silent = TRUE, was_forced = FALSE, visible = FALSE, update_icon = TRUE)
+	if(!IS_COMBAT_ACTIVE(src))
+		return
+	DISABLE_BITFIELD(combat_flags, COMBAT_FLAG_COMBAT_ACTIVE)
+	SEND_SIGNAL(src, COMSIG_LIVING_COMBAT_DISABLED, was_forced)
+	if(visible)
+		visible_message("<span class='warning'>[src] goes limp.</span>", "<span class='warning'>Your muscles are forceibly relaxed!</span>")
+	else if(!silent)
+		to_chat(src, was_forced? "<span class='warning'>Your muscles are forceibly relaxed!</span>" : "<span class='warning'>You relax your muscles.</span>")
```
Comment (Ghommie) [inline code/modules/mob/living/living_combat.dm]:
```suggestion
		to_chat(src, was_forced? "<span class='warning'>Your muscles are forcibly relaxed!</span>" : "<span class='warning'>You relax your muscles.</span>")
```

File: code/modules/mob/living/living_combat.dm [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+			enable_combat_mode(FALSE, TRUE, FALSE, FALSE)
+	update_combat_mode_icon()
+
+/mob/living/proc/disable_combat_mode(silent = TRUE, was_forced = FALSE, visible = FALSE, update_icon = TRUE)
+	if(!IS_COMBAT_ACTIVE(src))
+		return
+	DISABLE_BITFIELD(combat_flags, COMBAT_FLAG_COMBAT_ACTIVE)
+	SEND_SIGNAL(src, COMSIG_LIVING_COMBAT_DISABLED, was_forced)
+	if(visible)
+		visible_message("<span class='warning'>[src] goes limp.</span>", "<span class='warning'>Your muscles are forceibly relaxed!</span>")
```
Comment (Ghommie) [inline code/modules/mob/living/living_combat.dm]:
```suggestion
		visible_message("<span class='warning'>[src] goes limp.</span>", "<span class='warning'>Your muscles are forcibly relaxed!</span>")
```

File: code/modules/mob/living/living_combat.dm [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+/mob/living/proc/enable_combat_mode(silent = TRUE, was_forced = FALSE, visible = FALSE, update_icon = TRUE)
+	if(IS_COMBAT_ACTIVE(src))
+		return
+	ENABLE_BITFIELD(combat_flags, COMBAT_FLAG_COMBAT_ACTIVE)
+	SEND_SIGNAL(src, COMSIG_LIVING_COMBAT_ENABLED, was_forced)
+	if(visible)
+		visible_message("<span class='warning'>[src] drops into a combative stance!</span>", "<span class='warning'>You drop into a combative stance!</span>")
+	else if(!silent)
+		to_chat(src, was_forced? "<span class='warning'>Your muscles reflexively tignten!</span>" : "<span class='warning'>You tighten your muscles.</span>")
```
Comment (Ghommie) [inline code/modules/mob/living/living_combat.dm]:
```suggestion
		to_chat(src, was_forced? "<span class='warning'>Your muscles reflexively tighten!</span>" : "<span class='warning'>You tighten your muscles.</span>")
```

File: code/modules/mob/living/carbon/human/human_movement.dm [hunk 8]
```diff
@@ -9,6 +9,13 @@
 
 /mob/living/carbon/human/movement_delay()
 	. = ..()
+	if(CHECK_MOBILITY(src, MOBILITY_STAND) && m_intent == MOVE_INTENT_RUN && IS_SPRINTING(src))
```
Comment (Ghommie) [inline code/modules/mob/living/carbon/human/human_movement.dm]: One day these should be converted into movespeed modifiers, just so you know.

File: code/modules/mob/living/status_procs.dm [hunk 9]
```diff
@@ -283,13 +283,67 @@
...
+		if(I)
+			I.duration = max(world.time + amount, I.duration)
+		else if(amount > 0)
+			I = apply_status_effect(STATUS_EFFECT_STAGGERED, amount, updating)
+		return I
+
+/mob/living/proc/SetStaggered(amount, updating = TRUE, ignore_canstun = FALSE) //Sets remaining duration
+	if(SEND_SIGNAL(src, COMSIG_LIVING_STATUS_STAGGER, amount, updating, ignore_canstun) & COMPONENT_NO_STUN)
+		return
+	if(((status_flags & CANKNOCKDOWN) && !HAS_TRAIT(src, TRAIT_STUNIMMUNE)) || ignore_canstun)
```
Comment (Ghommie) [inline code/modules/mob/living/status_procs.dm]: early returns.

File: code/modules/mob/living/status_procs.dm [hunk 10]
```diff
@@ -283,13 +283,67 @@
...
+/mob/living/proc/AmountStaggered() //How many deciseconds remain in our Staggered status effect
+	var/datum/status_effect/staggered/I = IsStaggered()
+	if(I)
+		return I.duration - world.time
+	return 0
+
+/mob/living/proc/Stagger(amount, updating = TRUE, ignore_canstun = FALSE) //Can't go below remaining duration
+	if(SEND_SIGNAL(src, COMSIG_LIVING_STATUS_STAGGER, amount, updating, ignore_canstun) & COMPONENT_NO_STUN)
+		return
+	if(((status_flags & CANKNOCKDOWN) && !HAS_TRAIT(src, TRAIT_STUNIMMUNE)) || ignore_canstun)
```
Comment (Ghommie) [inline code/modules/mob/living/status_procs.dm]: idem.

File: code/modules/mob/living/status_procs.dm [hunk 11]
```diff
@@ -283,13 +283,67 @@
...
+			if(I)
+				I.duration = world.time + amount
+			else
+				I = apply_status_effect(STATUS_EFFECT_STAGGERED, amount, updating)
+		return I
+
+/mob/living/proc/AdjustStaggered(amount, updating = TRUE, ignore_canstun = FALSE) //Adds to remaining duration
+	if(SEND_SIGNAL(src, COMSIG_LIVING_STATUS_STAGGER, amount, updating, ignore_canstun) & COMPONENT_NO_STUN)
+		return
+	if(((status_flags & CANKNOCKDOWN) && !HAS_TRAIT(src, TRAIT_STUNIMMUNE)) || ignore_canstun)
```
Comment (Ghommie) [inline code/modules/mob/living/status_procs.dm]: .

**Review by Ghommie — CHANGES_REQUESTED** [review]
Fine. The combat/sprint locked sprites are a bit too imperceptible imo, you should make the figures/text flashier, maybe of a different hue too.

File: code/modules/mob/living/carbon/human/human_movement.dm (hunk 8 above)
Comment (silicons) [reply code/modules/mob/living/carbon/human/human_movement.dm]: yep i am just delaying the inevitable out of laziness :weary:

**Review by Ghommie — APPROVED** [review]

**Comment by Ghommie** [comment]
Travis compiled on the last commit but the status didn't update.

**Pull request merged**